### PR TITLE
GC: generic assay implementation

### PR DESCRIPTION
**Pull request by dippindots**
Fix cBioPortal/cbioportal#7654

Feature documentation:
https://docs.google.com/document/d/1_TuYpRICb6ru7hDiqhm0U0VI3ZSK93aa7tvb-pF5Z6A/edit?usp=sharing

**Testing:**

result view page (comparison tab):
https://deploy-preview-3351--cbioportalfrontend.netlify.app/results/comparison?Action=Submit&cancer_study_list=cesc_tcga_pan_can_atlas_2018&case_set_id=cesc_tcga_pan_can_atlas_2018_all&comparison_subtab=generic_assay_microbiome_signature&gene_list=TP53%250AEGFR&tab_index=tab_visualize
image

comparison page:
https://deploy-preview-3351--cbioportalfrontend.netlify.app/comparison/generic_assay_microbiome_signature?sessionId=5f4e85a5e4b070725d7dd237
image

### CODE REVIEW

**Comment by dippindots** [comment]
@cBioPortal/product Can I get your review on this new Generic Assay feature? We are trying to add one enrichment tab for each type of generic assay data (e.g. `Treatment Responses`, `Mutational Signatures`, `Microbiome`). Here are some screenshots: https://docs.google.com/document/d/1_TuYpRICb6ru7hDiqhm0U0VI3ZSK93aa7tvb-pF5Z6A/edit?usp=sharing.

**Comment by jjgao** [comment]
@dippindots the screenshots look good!

**Comment by kalletlak** [comment]
@dippindots looks like profile type(dataset) is appended to `GENERIC_ASSAY`, any specific reason? I think it is better to have it in its own field.

@alisman @jjgao @adamabeshouse any thoughts?

**Comment by adamabeshouse** [comment]
@kalletlak not sure I'm following, could you point to the relevant code?

File: src/pages/groupComparison/GenericAssayEnrichments.tsx
Comment (dippindots) [inline src/pages/groupComparison/GenericAssayEnrichments.tsx:99]: > @dippindots looks like profile type(dataset) is appended to `GENERIC_ASSAY`, any specific reason? I think it is better to have it in its own field.
> 
> @alisman @jjgao @adamabeshouse any thoughts?

@kalletlak Do you mean this `GenericAssayEnrichmentsDataSetDropdown`? This data set component is specific to `Generic Assay` since I cannot pass a Promise as the source of `dataSets`. I need to know which `generic assay type` is showing. That's why I created this new component.

Comment (kalletlak) [reply src/pages/groupComparison/GenericAssayEnrichments.tsx:99]: It is about `comparison_subtab` param in url. It is a join of `GENERIC_ASSAY` and dataset, example `GENERIC_ASSAY_MICROBIOME_SIGNATURE`. Instead we could have `comparison_subtab` as `generic_assay` and `MICROBIOME_SIGNATURE` as new parameter `dataset`. It would be nice if we standardize this even for other comparison subtabs.

Comment (dippindots) [reply src/pages/groupComparison/GenericAssayEnrichments.tsx:99]: @kalletlak I see, it's a good idea to separate the datasets as a new parameter. However I don't think I should leave `Generic Assay` as the `comparison_subtab` id since we will create multiple `comparison_subtabs` (one for each `generic assay type`), and then can select `dataset` from each subtabs.
https://github.com/cBioPortal/cbioportal-frontend/pull/3351/files#diff-d27fe25dc0b90c52a1e119c2a588de84R247-R276

File: src/pages/resultsView/enrichments/GenericAssayEnrichmentsBoxPlot.tsx
Comment (kalletlak) [inline src/pages/resultsView/enrichments/GenericAssayEnrichmentsBoxPlot.tsx:1]: How different is this from `ExpressionEnrichmentsBoxPlot.tsx`? can we combine them?

File: src/pages/resultsView/enrichments/GenericAssayEnrichmentsBoxPlot.tsx
Comment (kalletlak) [inline src/pages/resultsView/enrichments/GenericAssayEnrichmentsBoxPlot.tsx:200]: just the value is show. do generic assay type need to be shown?

File: src/pages/resultsView/enrichments/GenericAssayEnrichmentsDataSetDropdown.tsx
Comment (kalletlak) [inline src/pages/resultsView/enrichments/GenericAssayEnrichmentsDataSetDropdown.tsx:1]: This too, looks very much similar to `EnrichmentsDataSetDropdown.tsx`. combine them?

**Review by kalletlak — COMMENTED** [review]
Most of the code looks similar to expression enrichments. Is is possible to reuse existing code?

File: src/pages/resultsView/enrichments/MiniScatterChart.tsx
Comment (dippindots) [inline src/pages/resultsView/enrichments/MiniScatterChart.tsx:61]: Same as `ExpressionEnrichmentsBoxPlot.tsx`, we can use `onGenericAssayEntityClick` to know if this is a plot for Generic Assay data.

File: src/routes.jsx
Comment (dippindots) [inline src/routes.jsx:85]: A valid `Generic Assay` tab would starts with `GroupComparisonTab.GENERIC_ASSAY`

**Comment by dippindots** [comment]
@kalletlak @alisman @adamabeshouse I updated this pr with some comments. Can I get your review again? Thanks!

File: src/pages/groupComparison/GenericAssayEnrichments.tsx
Comment (adamabeshouse) [inline src/pages/groupComparison/GenericAssayEnrichments.tsx:74]: very nitpicky but in that case the code would be more readable if it was `const studyId = Object.keys(...)[0]` (and leave the comment as well)

File: src/pages/groupComparison/MethylationEnrichments.tsx
Comment (adamabeshouse) [inline src/pages/groupComparison/MethylationEnrichments.tsx:47]: why?

File: src/pages/groupComparison/MRNAEnrichments.tsx
Comment (adamabeshouse) [inline src/pages/groupComparison/MRNAEnrichments.tsx:48]: why?

File: src/pages/groupComparison/MutationEnrichments.tsx
Comment (adamabeshouse) [inline src/pages/groupComparison/MutationEnrichments.tsx:48]: why?

File: src/pages/groupComparison/ProteinEnrichments.tsx
Comment (adamabeshouse) [inline src/pages/groupComparison/ProteinEnrichments.tsx:47]: here too

File: src/pages/resultsView/ResultsViewPageHelpers.ts
Comment (adamabeshouse) [inline src/pages/resultsView/ResultsViewPageHelpers.ts:34]: caps when the others are lowercase?

File: src/pages/resultsView/enrichments/EnrichmentsDataSetDropdown.tsx
Comment (adamabeshouse) [inline src/pages/resultsView/enrichments/EnrichmentsDataSetDropdown.tsx:68]: why did you remove these two computeds? seems like it will lead to unnecessary recomputation

File: src/pages/resultsView/enrichments/ExpressionEnrichmentsBoxPlot.tsx
Comment (adamabeshouse) [inline src/pages/resultsView/enrichments/ExpressionEnrichmentsBoxPlot.tsx:154]: typo

File: src/pages/resultsView/enrichments/ExpressionEnrichmentsBoxPlot.tsx
Comment (adamabeshouse) [inline src/pages/resultsView/enrichments/ExpressionEnrichmentsBoxPlot.tsx:175]: can you add a comment to explain the regex

File: src/pages/resultsView/enrichments/ExpressionEnrichmentsBoxPlot.tsx
Comment (adamabeshouse) [inline src/pages/resultsView/enrichments/ExpressionEnrichmentsBoxPlot.tsx:186]: typo - molecular

File: src/shared/lib/comparison/ComparisonStore.ts
Comment (adamabeshouse) [inline src/shared/lib/comparison/ComparisonStore.ts:1189]: these names are too long to be readable. how about `gaGroupsByAssayType`?

File: src/shared/lib/comparison/ComparisonStore.ts
Comment (adamabeshouse) [inline src/shared/lib/comparison/ComparisonStore.ts:1242]: `gaDataGroupedByAssayType`? I am not sure what this is

File: src/shared/lib/comparison/ComparisonStore.ts
Comment (adamabeshouse) [reply src/shared/lib/comparison/ComparisonStore.ts:1189]: the names are very similar as well, which makes them harder to understand. could you please explain what this is so I can make a name recommendation?

File: src/shared/lib/comparison/ComparisonStore.ts
Comment (adamabeshouse) [reply src/shared/lib/comparison/ComparisonStore.ts:1242]: can you please explain what this is?

File: src/shared/lib/comparison/ComparisonStore.ts
Comment (adamabeshouse) [inline src/shared/lib/comparison/ComparisonStore.ts:1285]: can you please explain what this is?

File: src/pages/resultsView/enrichments/GenericAssayEnrichmentsTableDataStore.ts
Comment (adamabeshouse) [inline src/pages/resultsView/enrichments/GenericAssayEnrichmentsTableDataStore.ts:4]: any?

File: src/pages/groupComparison/MRNAEnrichments.tsx
Comment (kalletlak) [reply src/pages/groupComparison/MRNAEnrichments.tsx:48]: I think it should be `mRNAEnrichmentProfiles`

Comment (dippindots) [reply src/pages/groupComparison/MRNAEnrichments.tsx:48]: What did I do.... lost my mind...

File: src/pages/resultsView/ResultsViewPageHelpers.ts
Comment (dippindots) [reply src/pages/resultsView/ResultsViewPageHelpers.ts:34]: I can change it to lower case

File: src/pages/resultsView/enrichments/EnrichmentsDataSetDropdown.tsx
Comment (dippindots) [reply src/pages/resultsView/enrichments/EnrichmentsDataSetDropdown.tsx:68]: The reason I removed those two computes is: The `this.props.dataSets` is not a promise anymore, so when we deal with them, we don't need to see if the dataSets is loading completed. They may don't need to be recomputed. I moved the logic below from L58-L65

File: src/pages/resultsView/enrichments/GenericAssayEnrichmentsTableDataStore.ts
Comment (dippindots) [reply src/pages/resultsView/enrichments/GenericAssayEnrichmentsTableDataStore.ts:4]: updated to `GenericAssayEnrichmentRow`

File: src/shared/lib/comparison/ComparisonStore.ts
Comment (dippindots) [reply src/shared/lib/comparison/ComparisonStore.ts:1189]: It's similar to `mrnaEnrichmentAnalysisGroups`, it prepares the groups to be sent to backend for data fetching.

Some background of generic assay:
`Generic Assay` profiles can have different `genericAssayType`: https://github.com/cBioPortal/cbioportal/blob/master/db-scripts/src/main/resources/cgds.sql#L317.

In one study, we may have many profiles with different `genericAssayType`, so we want to generate one group comparison sub-tab for each `genericAssayType`. Which means we need to create a map (keyed by `genericAssayType`) to store data.

Comment (dippindots) [reply src/shared/lib/comparison/ComparisonStore.ts:1189]: Maybe `gaEnrichmentAnalysisGroupsMap`?

File: src/shared/lib/comparison/ComparisonStore.ts
Comment (dippindots) [reply src/shared/lib/comparison/ComparisonStore.ts:1242]: This is another map (keyed by `genericAssayType`), which prepares `group` parameter to be sent to `fetchGenericAssayEnrichmentsUsingPOST` API.

File: src/pages/groupComparison/GroupComparisonPage.tsx
Comment (alisman) [inline src/pages/groupComparison/GroupComparisonPage.tsx:222]: no reason to nest inside of lodash map here.  _.keys returns array so you can just chain _.keys(..).map(..)

File: src/pages/groupComparison/MRNAEnrichments.tsx
Comment (alisman) [reply src/pages/groupComparison/MRNAEnrichments.tsx:48]: good catch @adamabeshouse !

File: src/shared/lib/comparison/ComparisonStore.ts
Comment (dippindots) [reply src/shared/lib/comparison/ComparisonStore.ts:1285]: This function get the enrichment data map (keyed by `genericAssayType`). After we have this, we can use `genericAssayType` to get data for each comparison sub-tab.

File: src/pages/resultsView/ResultsViewPageStoreUtils.ts
Comment (alisman) [inline src/pages/resultsView/ResultsViewPageStoreUtils.ts:725]: i wouldn't use turnary this way.  it's confusing.  just use if/else.

File: src/pages/resultsView/ResultsViewPageStoreUtils.ts
Comment (alisman) [inline src/pages/resultsView/ResultsViewPageStoreUtils.ts:847]: my guess is this could be typed?   i.e. pValue and stableId are members of some type?

File: src/pages/resultsView/comparison/ComparisonTab.tsx
Comment (alisman) [inline src/pages/resultsView/comparison/ComparisonTab.tsx:248]: chain instead of nest

File: src/pages/resultsView/enrichments/AlterationEnrichmentsContainer.tsx
Comment (alisman) [inline src/pages/resultsView/enrichments/AlterationEnrichmentsContainer.tsx:100]: should this be named filterByGene?  i mean, it's not generic so why not be more specific

File: src/pages/resultsView/enrichments/EnrichmentsDataSetDropdown.tsx
Comment (alisman) [inline src/pages/resultsView/enrichments/EnrichmentsDataSetDropdown.tsx:62]: use _.some which will return boolean

File: src/pages/resultsView/enrichments/EnrichmentsDataSetDropdown.tsx
Comment (adamabeshouse) [reply src/pages/resultsView/enrichments/EnrichmentsDataSetDropdown.tsx:68]: I see, yes but the @computed is not just about checking the promise status, it's also about saving on recomputation as long as dataSets doesn't change. the way you have it now, it will recompute in every render which is unnecessary

File: src/shared/lib/comparison/ComparisonStore.ts
Comment (adamabeshouse) [reply src/shared/lib/comparison/ComparisonStore.ts:1189]: oh okay, thanks. I think that the word 'analysis' may not be necessary. I still think its nice to keep "by assay type". how about `gaEnrichmentGroupsByAssayType`?

File: src/shared/lib/comparison/ComparisonStore.ts
Comment (adamabeshouse) [reply src/shared/lib/comparison/ComparisonStore.ts:1242]: thanks. how about `gaDataQueriesByAssayType`?

File: src/shared/lib/comparison/ComparisonStore.ts
Comment (adamabeshouse) [reply src/shared/lib/comparison/ComparisonStore.ts:1285]: thanks. how about `gaEnrichmentDataByAssayType`?

File: src/shared/lib/comparison/ComparisonStore.ts
Comment (dippindots) [reply src/shared/lib/comparison/ComparisonStore.ts:1189]: Sounds good, I will rename them!

File: src/shared/lib/comparison/ComparisonStore.ts
Comment (dippindots) [reply src/shared/lib/comparison/ComparisonStore.ts:1242]: Sounds good, I will rename them!

File: src/shared/lib/comparison/ComparisonStore.ts
Comment (dippindots) [reply src/shared/lib/comparison/ComparisonStore.ts:1285]: Yes!

File: src/pages/resultsView/enrichments/EnrichmentsDataSetDropdown.tsx
Comment (dippindots) [reply src/pages/resultsView/enrichments/EnrichmentsDataSetDropdown.tsx:68]: @adamabeshouse Thanks, that's right. I will bring the compute back.

File: src/pages/resultsView/ResultsViewPageStoreUtils.ts
Comment (dippindots) [reply src/pages/resultsView/ResultsViewPageStoreUtils.ts:847]: Yes, it can be typed. I will give it the type `GenericAssayEnrichmentWithQ` to it.

File: src/pages/resultsView/ResultsViewPageStoreUtils.ts
Comment (dippindots) [reply src/pages/resultsView/ResultsViewPageStoreUtils.ts:725]: Sure

**Review by adamabeshouse — APPROVED** [review]

**Review by kalletlak — APPROVED** [review]
LGTM! 

Squash commits.

**Comment by dippindots** [comment]
@kalletlak @adamabeshouse @alisman Thanks for all the help with reviewing this pr! I will wait until the tests passing and then merge it.

**Review by alisman — APPROVED** [review]

**Pull request merged**
